README
Perl library for controlling UAVs, specifically ARDrone (maybe others in the future).  See:

http://ardrone2.parrot.com

WARNING!
If your ARDrone crashes into your neighbor's house, invades Iranian airspace, or tries to 
find Sarah Conner, don't blame me.


RECOMMENDED MODULES
===================

SDL is highly recommended on clients for displaying navigation data, video 
streams, and joystick input.

The RaspberryPi HiPi modules are only needed for the WumpusRover server.


INSTALLATION
============

For full functionality, you will need the SDL Perl module and the ffmpeg C library.

Install with the standard Module::Build method:

    perl ./Build.PL
    ./Build
    ./Build test
    ./Build install



SPECIAL THANKS
==============

Nirmal Biswas for font 'Typeone': http://www.dafont.com/nirmal-biswas.d3277

Curtis Brandt for helping get things working on Macs


CHANGELOG
Revision history for perl module UAV::Pilot

1.2 2015-08-22

- Install exec scripts under Dist::Zilla

1.1 2015-06-08

- Track down issues with version number
- Bump copyright year

1.0 2014-06-05

- Bump version number.  Really, that's it.

1.0_1 2014-06-04

- Allow loading of other modules in Commands packages

1.0_0 2014-06-04

- No longer use share/ method of loading Command shell modules.  Instead, 
  use a specific namespace with some @INC magic.

0.11 2014-05-29

- Port build to Dist::Zilla
- Actually fix regression bug of FileDump taking a filehandle

0.10 2014-04-27

- Fix regression bug of FileDump taking a filehandle
- UAV::Pilot::Command calls uav_module_quit() when it's about to exit

0.9 2014-02-16

- Breakoff parts into new dists

0.8 2013-12-28

- Unicast implemented for Parrot AR.Drone navdata
- WumpusRover implemented
- Decoupled interface in bin/uav and share/*.uav libaries to support more than 
  just the AR.Drone
- Decoupled SDL Joystick interface from AR.Drone
- UAV::Pilot::SDL::JoystickConverter deleted as unnecessary
- Change ARDrone::Control::Event interface to take an existing condvar rather 
  than creating and returning a new one
- Fixed a crash where the AR.Drone's navdata was sending NaN floats

0.7 2013-10-24

- Cleanup a bunch of bad tests and dependencies from 0.6

0.6 2013-10-15

[EventHandler Interface]
- UAV::Pilot::SDL::EventHandler moved to UAV::Pilot::EventHandler
- UAV::Pilot::Events based on previous UAV::Pilot::SDL::Events, except it 
  doesn't do anything for SDL
- UAV::Pilot::SDL::Events changed to an EventHandler
- All other EventHandler objects and bin/uav* scripts updated to reflect this 
  change

[Controller and Driver Interface]
- Rename UAV::Pilot::Driver::ARDrone to UAV::Pilot::ARDrone::Driver
- Rename UAV::Pilot::Control::ARDrone to UAV::Pilot::ARDrone::Control
- All the other modules under those namespaces were renamed accordingly

[Other Changes]
- In UAV::Pilot::Control, change the 'sender' param to 'driver'
- Stack handlers passed to UAV::Pilot::Video::H264Decoder.  Note the API change 
  of 'display' to 'displays'
- Stack handlers passed to UAV::Pilot::Driver::ARDrone::Video.  Note the API 
  change of 'handler' to 'handlers'
- 'uav' shell handles dumping video to file while displaying it at the same time
- Implement 'NavCollector' to receive nav packets as they're read from the driver
- SDLNavDisplay implemented as a NavCollector rather than an EventHandler
- When we get a bad PaVE header on the video stream, fast forward to the next
  one.  This helps recover the video stream gracefully.  The video will be 
  messed up until the next keyframe, but at least it keeps going.


0.5 2013-07-31

- Real-time video implemented (this is the big one)
- VideoHandler and FileDump moved into UAV::Pilot::Video namespace.  Handler 
  role now named "H264Handler".
- Joystick buttons can be mapped to actions
- Joystick module can take a path to a conf file
- bin/uav gets an --iface option for a network interface for binding


0.4 2013-07-04

- Dump video to file
    - If using mplayer to show the file, set the FPS manually with 
      "mplayer -fps 30 /path/to/file.h264"
- Rename anything that starts an event loop to init_event_loop()
- Joystick to UAV control number conversion decoupled from ARDrone specifics
- Add methods to UAV::Pilot::Control::ARDrone for userbox nav data and picture 
  taking


0.3 2013-06-17

- Event-based interface
- Joystick control


0.2 2013-06-02

- Device renamed to Control
- Sender renamed to Driver
- Fetch demo navigation data


0.1 2013-05-09

- Piloting commands work


scripts/video_dump.pl
#!/usr/bin/perl
use v5.14;
use warnings;
use IO::Socket::INET;
use Data::Dumper 'Dumper';

local $Data::Dumper::Sortkeys = 1;

use constant PAVE_SIGNATURE => 'PaVE';
use constant BUF_READ_SIZE  => 4096;

my $HOST = '192.168.1.1';
my $PORT = 5555;
my $FILE = shift;


my $num_frames = 0;
sub read_frame
{
    my ($input) = @_;
    my $buf;
    my $got_input = $input->read( $buf, BUF_READ_SIZE );
    return {}, 0 if ! $got_input;

    my @bytes = unpack "C*", $buf;

    my %packet;
    $packet{signature}               = pack 'c4', @bytes[0..3];
    $packet{version}                 = $bytes[4];
    $packet{video_codec}             = $bytes[5];
    $packet{packet_size}             = convert_16bit_LE( @bytes[6,7]);
    $packet{payload_size}            = convert_32bit_LE( @bytes[8..11] );
    $packet{encoded_stream_width}    = convert_16bit_LE( @bytes[12,13] );
    $packet{encoded_stream_height}   = convert_16bit_LE( @bytes[14,15] ); # here
    $packet{display_width}           = convert_16bit_LE( @bytes[16,17] );
    $packet{display_height}          = convert_16bit_LE( @bytes[18,19] );
    $packet{frame_number}            = convert_32bit_LE( @bytes[20..23] );
    $packet{timestamp}               = convert_32bit_LE( @bytes[23..26] );
    $packet{total_chunks}            = $bytes[27];
    $packet{chunk_index}             = $bytes[28];
    $packet{frame_type}              = pack 'C', $bytes[29];
    $packet{control}                 = $bytes[30];
    $packet{stream_byte_position_lw} = convert_32bit_LE( @bytes[31..34] );
    $packet{stream_byte_position_uw} = convert_32bit_LE( @bytes[35..38] );
    $packet{stream_id}               = convert_16bit_LE( @bytes[39,40] );
    $packet{total_slices}            = $bytes[41];
    $packet{slice_index}             = $bytes[42];
    $packet{packet1_size}            = $bytes[43];
    $packet{packet2_size}            = $bytes[44];
    $packet{reserved2}               = pack 'C2', @bytes[45,46];
    $packet{advertised_size}         = convert_32bit_LE( @bytes[47..50] );
    $packet{reserved3}               = pack 'C12', @bytes[51..62];
    warn "Bad PaVE signature, got: '$packet{signature}'\n"
        if PAVE_SIGNATURE ne $packet{signature};

    my ($payload, $continue_reading) = read_frame_payload(
        [@bytes[$packet{packet_size}..$#bytes]] ,
        $input,
        $packet{payload_size}
    );
    $packet{payload} = $payload;
    $num_frames++;
    return (\%packet, $continue_reading);
}

sub read_frame_payload
{
    my ($leftover_bytes, $input, $total_size) = @_;
    my @bytes = @$leftover_bytes;

    my $continue = 1;
    while( (scalar(@bytes) < $total_size) && $continue ) {
        my $buf;
        my $size_left = $total_size - scalar(@bytes);
        my $buf_size = ($size_left > BUF_READ_SIZE)
            ? BUF_READ_SIZE
            : $size_left;
        my $bytes_recv = $input->read( $buf, $buf_size );
        $continue = 0 if ! $bytes_recv;

        push @bytes, unpack( 'C*', $buf );
    }

    return (\@bytes, $continue);
}


sub convert_32bit_LE
{
    my (@bytes) = @_;
    my $val = $bytes[0]
        | ($bytes[1] << 8)
        | ($bytes[2] << 16)
        | ($bytes[3] << 24);
    return $val;
}

sub convert_16bit_LE
{
    my (@bytes) = @_;
    my $val = $bytes[0] | ($bytes[1] << 8);
    return $val;
}


{
    my $input = undef;
    if( defined $FILE) {
        open( $input, '<', $FILE ) 
            or die "Can't open file for reading: $!\n";
    }
    else {
        $input = IO::Socket::INET->new(
            PeerAddr  => $HOST,
            PeerPort  => $PORT,
        ) or die "Can't open socket to $HOST:$PORT: $!\n";
    }

    my $continue = 1;
    while( $continue ) {
        my $frame = {};
        ($frame, $continue) = read_frame( $input );
        print pack( 'C*', @{ $$frame{payload} } ) if %$frame;
    }


    $input->close;
    warn "Frames processed: $num_frames\n";
}
